src/Microsoft.Health.Fhir.CosmosDb/Features/Search/FhirCosmosSearchService.cs

```diff
@@ -449,23 +455,59 @@ private async Task<Expression> RecurseChainedExpression(ChainedExpression expres
                     searchEnumerationTimeoutOverrideIfSequential = TimeSpan.FromSeconds(5);
 
                     // Executing query sequentially until the timeout
-                    (results, nextContinuationToken) = await _fhirDataStore.ExecuteDocumentQueryAsync<T>(sqlQuerySpec, feedOptions, null, searchOptions.MaxItemCountSpecifiedByClient, searchEnumerationTimeoutOverrideIfSequential, cancellationToken);
+                    (results, nextContinuationToken) = await _fhirDataStore.ExecuteDocumentQueryAsync<T>(
+                        sqlQuerySpec: sqlQuerySpec,
+                        feedOptions: feedOptions,
+                        mustNotExceedMaxItemCount: searchOptions.MaxItemCountSpecifiedByClient,
+                        searchEnumerationTimeoutOverride: searchEnumerationTimeoutOverrideIfSequential,
+                        cancellationToken: cancellationToken);
 
                     // check if we need to restart the query in parallel
                     if (results.Count < desiredItemCount && !string.IsNullOrEmpty(nextContinuationToken))
                     {
                         feedOptions.MaxConcurrency = _cosmosConfig.ParallelQueryOptions.MaxQueryConcurrency;
-                        (results, nextContinuationToken) = await _fhirDataStore.ExecuteDocumentQueryAsync<T>(sqlQuerySpec, feedOptions, null, searchOptions.MaxItemCountSpecifiedByClient, null, cancellationToken);
+                        (results, nextContinuationToken) = await _fhirDataStore.ExecuteDocumentQueryAsync<T>(
+                            sqlQuerySpec: sqlQuerySpec,
+                            feedOptions: feedOptions,
+                            mustNotExceedMaxItemCount: searchOptions.MaxItemCountSpecifiedByClient,
+                            cancellationToken: cancellationToken);
                     }
                 }
                 else
                 {
-                    (results, nextContinuationToken) = await _fhirDataStore.ExecuteDocumentQueryAsync<T>(sqlQuerySpec, feedOptions, continuationToken, searchOptions.MaxItemCountSpecifiedByClient, feedOptions.MaxConcurrency == null ? searchEnumerationTimeoutOverrideIfSequential : null, cancellationToken);
+                    var mustNotExceedMaxItemCount = searchOptions.MaxItemCountSpecifiedByClient;
+                    FeedRange queryFeedRange = null;
+
+                    // Large async operations also need the full result set. Accounting for the till factor to achieve that.
+                    if (searchOptions.IsLargeAsyncOperation)
+                    {
+                        mustNotExceedMaxItemCount = false;
+                        feedOptions.MaxItemCount = (int)(feedOptions.MaxItemCount / CosmosFhirDataStore.ExecuteDocumentQueryAsyncMinimumFillFactor);
+
+                        try
+                        {
+                            queryFeedRange = searchOptions.FeedRange is not null ? FeedRange.FromJsonString(searchOptions.FeedRange) : queryFeedRange;
```

Comment:
Nit: Setting queryFeedRange equal to itself when the only possible value at this point is null seems unnecessarily confusing. I think just setting it equal to null would be clearer.

Reply:
Agreed - thanks. Sometimes one liners are not the right approach, haha.